Mcalus3/add catalog import plugin

mcalus3 opened the pull request:
## Feedback request

I would like to have some feedback about adding entity config generator to the existing codebase of catalog-backend. I tried to mimic how "HigherOrderOperations" class is being used, but I don't know if it's the good way to have it. The function that generates the config file right now has hardcoded only simple string, but in the future it should be able to use multiple processors for it.

Link to video of the frontend component:
https://www.loom.com/share/318d68452c4549539751b00356d42ab8

stefanalund commented on packages/app/package.json:
```diff
@@ -37,6 +37,7 @@
     "@roadiehq/backstage-plugin-github-insights": "^0.2.7",
     "@roadiehq/backstage-plugin-github-pull-requests": "^0.5.2",
     "@roadiehq/backstage-plugin-travis-ci": "^0.2.3",
+    "@roadiehq/backstage-plugin-catalog-import": "^0.1.0",
```
As we discussed, since this will be part of the core experience it needs to be part of the main repo. Ping @dtuite

dtuite replied:
Hey Stefan. I dropped off the call before this was discussed. I will send you a message on Discord.

mcalus3 commented:
@freben Would you like to give a feedback on how you would like to have this code structured? Right now I've just created a new class similar to `HigherOrderOperations` that is being called in the route handler (i though that this functionality could potentially grow independently from the `HigherOrderOperations`, but maybe it's not worth having one more object being passed around in the builders?) I also can't find any satisfying option for using existing processors. Do you think tthat there should be created a new type of location ("repo" type to indicate that location is pointing o the repository instead of the config file) and invoke a `locationReader.read()` on it? Or should I just inject all the processors to my new `ConfigGenerator`class and iterate over those of them that have `preProcessEntity` and `postProcessEntity` methods?

And would it be possible to get approved and merged to master such a "minimum viable" version first that would only generate a bare entity (with name and owner taken from the repo path)?

dtuite commented:
This PR contains:

 1. A backend `ConfigGenerator` which can generate a YAML config representing an entity and send it back to the client, so that it can subsequently open a PR against the repo.
 2. A backend API to accept the URL of a repo and run it through the `ConfigGenerator`.
 3. A Backstage plugin which contains:
       1. a UI for choosing GitHub repos and sending them to the backend API,
       2. Code which, when given the YAML config, can open a PR against the repo in question, to add a `catalog-info.yaml`.

Repo processors can come in an other PR, with more guidance from the core team.

The plugin with the UI and repo will be contributed directly to Backstage instead of living as a separate plugin. It's so close to the core Backstage experience that it just makes more sense to do that. Also the code for opening PRs could be applicable in other places.

freben commented on packages/catalog-model/src/location/validation.ts:
```diff
@@ -34,3 +34,10 @@ export const locationSchema = yup
   .noUnknown()
   .required();
+
+export const repoPathSchema = yup
```
I believe the input would be 

```ts
type AnalyzeLocationRequest = {
  location: LocationSpec;
};
```

for now; there may eventually be more fields added to that but at this point a location probably suffices. There is probably no reason to invent a separate way to represent the location to be analyzed.

The output could be something like

(This is all just suggestions as a basis for discussion!)

```ts
type AnalyzeLocationResponse = {
  existingEntityFiles: AnalyzeLocationExistingEntity[];
  generateEntities: AnalyzeLocationGenerateEntity[];
}

// If the folder pointed to already contained catalog info yaml files, they are
// read and emitted like this so that the frontend can inform the user that it
// located them and can make sure to register them as well if they weren't
// already
type AnalyzeLocationExistingEntity = {
  location: LocationSpec;
  isRegistered: boolean;
  entity: Entity;
};

// This is some form of representation of what the analyzer could deduce.
// We should probably have a chat about how this can best be conveyed to
// the frontend. It'll probably contain a (possibly incomplete) entity, plus
// enough info for the frontend to know what form data to show to the user
// for overriding/completing the info.
type AnalyzeLocationGenerateEntity = {
  // Some form of partial representation of the entity
  entity: RecursivePartial<Entity>;
  // Lists the suggestions that the user may want to override
  fields: AnalyzeLocationEntityField[];
};

// This is where I get really vague. Something like this perhaps? Or it could be
// something like a json-schema that contains enough info for the frontend to
// be able to present a form and explanations
type AnalyzeLocationEntityField = {
  // e.g. "spec.owner"? The frontend needs to know how to "inject" the field into the
  // entity again if the user wants to change it
  field: string;

  // The outcome of the analysis for this particular field
  state: 'analysis_suggested_value' | 'analysis_suggested_no_value' | 'needs_user_input';

  // If the analysis did suggest a value, this is where it would be. Not sure if we want
  // to limit this to strings or if we want it to be any JsonValue
  value: string | null;

  // A text to show to the user to inform about the choices made. Like, it could say
  // "Found a CODEOWNERS file that covers this target, so we suggest leaving this
  // field empty; which would currently make it owned by X" where X is taken from the
  // codeowners file.
  description: string; 
};
```

mcalus3 replied:
Added types that you suggested

freben commented on plugins/catalog-backend/src/ingestion/ConfigGenerator.ts:
```diff
@@ -0,0 +1,45 @@
+import { Logger } from 'winston';
+import { ConfigGenerator } from './types';
+
+export class ConfigGeneratorClient implements ConfigGenerator {
```
The word "config" is overloaded, usually referring to the actual config of the app. May want to use combinations of "location" and "analyze" in my opinion - `LocationAnalyzer`, with a method `analyzeLocation` that takes an `AnalyzeLocationRequest` etc

mcalus3 replied:
I agree that the word "Config" is too generic, but generating is *what* this feature will do, analyzing path is only the *how*. Other names I was considering was combinations of "Component|Entity", "Info|Config" and then "Generator|Creator". Maybe you would like to have this name more like that? (either way, you have better knowledge on how it interacts with other names in the codebase, so just choose it, I will adapt to it)

mcalus3 replied:
Changed to LocationAnalyzer

freben commented on plugins/catalog-backend/src/ingestion/ConfigGenerator.ts:
```diff
@@ -0,0 +1,45 @@
+import { ConfigGenerator } from './types';
+
+export class ConfigGeneratorClient implements ConfigGenerator {
+  logger: Logger;
```
`private readonly` :)

mcalus3 replied:
Changed

freben commented on plugins/catalog-backend/src/ingestion/types.ts:
```diff
@@ -65,3 +65,11 @@ export type ReadLocationError = {
+//
+
+export type ConfigGenerator = {
+  generateConfig(repoPath: string): Promise<Entity[]>;
```
yeah so probably renaming this according to the above comment, and taking something like an `AnalyzeLocationRequest` and returning `Promise<AnalyzeLocationResponse>`

mcalus3 replied:
Changed

freben commented on plugins/catalog-backend/src/service/CatalogBuilder.ts:
```diff
@@ -249,6 +251,7 @@ export class CatalogBuilder {
     entitiesCatalog: EntitiesCatalog;
     locationsCatalog: LocationsCatalog;
     higherOrderOperation: HigherOrderOperation;
+    configGenerator: ConfigGenerator;
```
I agree that this is turning into a bit of a mess of classes, if it keeps growing for each function that we introduce, but I think it's still the sensible choice at the moment

mcalus3 replied:
That was my intuition - we will even be able to stub it or block by some feature flags thanks to that separation.

freben commented on plugins/catalog-backend/src/ingestion/ConfigGenerator.ts:
```diff
@@ -0,0 +1,45 @@
+      kind: 'Component',
+      metadata: {
+        name: repoName,
+        description: '',
```
this is an optional field and can be left out entirely if we don't have something to put in it at the moment

mcalus3 replied:
Removed for now

freben commented on plugins/catalog-backend/src/ingestion/ConfigGenerator.ts:
```diff
@@ -0,0 +1,45 @@
+        description: '',
+        annotations: { 'github.com/project-slug': repoPath },
+      },
+      spec: { type: 'service', owner: ownerName, lifecycle: 'experimental' },
```
The `type` will probably be one of those fields that is a bit hard to deduce and that the user may want to fill in into a form (possibly as a dropdown) after this step completes.

Owner we can do something about when there are codeowners at least.

Lifecycle - yeah that will be hard to deduce as well, but maybe it's very common that the thing you import is prod (because it's "old", not generated through the Backstage templating), so this could default to `'production'`.

freben commented on plugins/catalog-backend/src/ingestion/ConfigGenerator.ts:
```diff
@@ -0,0 +1,45 @@
+      metadata: {
+        name: repoName,
+        description: '',
+        annotations: { 'github.com/project-slug': repoPath },
```
so this particular annotation can be added IF the location we pointed to was a github one, and IF we found a `.github/workflows` folder right? Or other things that are triggered by that annotation, if any

freben commented on plugins/catalog-backend/src/service/router.ts:
```diff
@@ -125,6 +131,14 @@ export async function createRouter(
   }
 
+  if (configGenerator) {
+    router.post('/generate-config', async (req, res) => {
```
probably `/analyze-location`

mcalus3 replied:
Changed

freben reviewed (commented):
Don't be scared by the amount of comments, we're a bit wavering ourselves on the exact shape of things in the end and had some open discussions that are reflected in the comments too. A lot of it is meant as inspiration or discussion pieces rather than pointing fingers. Keep up the good work and let us know if something seems off or unclear.

mcalus3 commented on packages/app/src/App.tsx:
```diff
@@ -83,6 +84,7 @@ const AppRoutes = () => (
       element={<RegisterComponentRouter catalogRouteRef={catalogRouteRef} />}
     />
     <Route path="/settings" element={<SettingsRouter />} />
+    <Route path="/import-catalog/*" element={<CatalogImportRouter />} />
```
I agree, but for some reason if the route is `catalog-*`, blank page loads instead of the pluginRouter component. That's why for now i've inverted this route. I think I will debug it at the end of the work.

mcalus3 commented on packages/app/src/apis.ts:
```diff
@@ -59,8 +49,4 @@ export const apis = [
-
-  // TODO: move to plugins
-  createApiFactory(travisCIApiRef, new TravisCIApi()),
-  createApiFactory(githubPullRequestsApiRef, new GithubPullRequestsClient()),
```
Now that https://github.com/spotify/backstage/pull/3085 is merged, it's unnecessary. It should disappear after merge to master.

freben replied:
You should totally rebase this branch on the new origin master though. These things left in a PR can turn into an actual problem if there are deviations that weren't caught

mcalus3 replied:
It's gone now after merge to upstream/master.

mcalus3 commented:
@freben I've created a frontend plugin for importing a single repo (video link in the PR description). Right now it's still a frankenstein monster created from the register-component plugin and our prototype that we were building for bulk imports. I will still be cleaning up there and will change name to "component-import" as we previously agrred on (i've set different name because I already have plugin with this name that waits to be renamed)

dtuite commented:
Here is a mockup of the UX we are going for. https://whimsical.com/github-importer-UiLo6FicL54kw2Do1KxDbY

Feedback welcome.

freben commented:
@dtuite Ah, that's very helpful! Looking great, very minor questions:

- Button text "Fetch" - hm, just "Next" perhaps? As the user doesn't so to speak need to decide on the technical underlying action here.
- Button text "Open pull request" - perhaps "Create" instead of "Open" to not collide with the notion of navigating to it with the browser?
- Some providers call them "Merge Requests" instead of "Pull Requests" right? I dunno if it's at all worth caring about the distinction, just pondering.

dtuite commented:
@freben 

 - 👍 
 - 👍 
 - Theoretically we could change the button text depending on which CVS hosting they use. I suggest we stick with Pull Request for now and a later PR could make it dynamic.

Your third point has made me realise that we may need the user to tell us which type of VCS hosting they use. An on-prem GitHub can have a custom hostname, I presume an on-prem GitLab can too. If we're going to fetch the repo we need to know the API to use.

dtuite commented:
Marek has been out recently which is why this isn't progressing. We should see some movement next week.

freben commented:
The backend already has enough info to know the provider type, and its API endpoint URL, thanks to the `integrations` config. If you want to have that returned in the initial analysis response, we can make it happen. May need to add a function or two first in the backend.

andrewthauer commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -0,0 +1,188 @@
+        method: 'POST',
+        body: JSON.stringify({
+          type: 'github',
+          target: `https://github.com/${owner}/${repo}/blob/master/catalog-info.yaml`,
```
Assuming the default branch is `master` is probably not a good assumption. It's also becoming more popular to use `main`. I'm wondering if the GH API can return what the default branch name is for a repo?

mcalus3 replied:
Yest it is, im' using it in the "submitPRToRepo" method. Will add it here also.

mcalus3 replied:
Fixed

andrewthauer commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -0,0 +1,188 @@
+      auth: token,
+    });
+
+    const branchName = 'backstage-integration';
```
Probably not a problem in most cases, but some organizations may have strict branch naming conventions and CI that might fail the PR. Probably not worth addressing in the initial version though.

mcalus3 replied:
In the future it's meant to be provided by the user from the UI and the _backstage-integration_ will remain just as a default value.

andrewthauer commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -0,0 +1,188 @@
+      .catch(e => {
+        throw new Error(
+          formatHttpErrorMessage(
+            `Couldn't create a commit with ${fileName} file added`,
```
"Could not create a commit. The file '${fileName}' already exists." Instead maybe?

mcalus3 replied:
It's generic message for all errors that would occur during an attempt to create a commit - I don't want to say it's because the file already exist (we are not sure), but user would be able to check if it's the case.

andrewthauer replied:
Something about the wording was throwing me off. Makes more sense now. It's easy enough to modify later if needed.

andrewthauer commented on plugins/catalog-import/src/components/ImportComponentForm.tsx:
```diff
@@ -0,0 +1,119 @@
+    try {
+      if (!isMounted()) return;
+
+      const repo = target.split('/').slice(-2).join('/');
```
Using git-url-parse might be safer to use here.

mcalus3 replied:
Fixed

andrewthauer commented on plugins/catalog-import/src/util/useGithubRepos.ts:
```diff
@@ -0,0 +1,63 @@
+  const auth = useApi(githubAuthApiRef);
+
+  const submitPRToRepo = async (selectedRepo: ConfigSpec) => {
+    const token = await auth.getAccessToken(['repo']);
```
I've been thinking about this. What are peoples thoughts about using either backend as a proxy with the GITHUB_TOKEN for scenarios where readonly GH API access is only required? Having to setup GH OAuth and also introducing multiple login providers (in some cases) seems to create a higher barrier of entry.

mcalus3 replied:
We want the frontend plugin to work in behalf of the user's github account to create pull requests as them.

andrewthauer replied:
That's a fair point for sure, but also creates another barrier. Using an OAuth token in this manner likely requires a GH org admin to be involved. So I could see this being more contentious and or time consuming to setup in some organizations. I'm not necessarily suggesting to change it here, but pointing out some downsides to the GH OAuth approach. I think approaching the API aspect of this plugin from a perspective where it "could" use a backend implemented API in the future would be ideal. That said that can be a future consideration as well.

freben replied:
We have this situation at Spotify. The login provider is a specialised one that decodes a Google IAP token and looks up user info in LDAP, so there's a Google login screen to get past for visiting the page, but it happens extremely rarely since the session is picked up and renewed silently with a very long expiry. Then when you want to perform some action that requires GitHub Enterprise access (like this PR), the FIRST time you will be asked for consent and then pretty much never again since we make use of refresh tokens. So the cost for the end user is one login screen and one consent popup extremely seldom. The cost for the integrator is indeed to need an OAuth application to be added to their GitHub (if that's what they use), but that's a one time cost. The token that the proxy or backend would use, also needs the GitHub owner's involvement to get, right?

andrewthauer replied:
@freben - I'm no expert, but based on my experimentation, the OAuth token and standard GH user tokens are not equivalent. We've been able to use a standard (non GH org admin) user token for the backend integrations without any problems. However, the only way to get plugins that use GH auth to work on private repos is to have a GH org admin create the OAuth token. All that said, I'm not hung up on this point, just having an open conversation 😄

Rugvip replied:
@andrewthauer Tricky bit would be to seamlessly transition between the two. Integrations work is heading in that direct along with `UrlReader`s etc. The possibility of a `RepoApi` is something we've surfaced a couple of times, which could make it a bit easier to work with repos without having to care exactly how authentication happens or where the repos reside. There's plenty more plumbing to get in place before we'd be able to reach something like that though, as far as I can tell.

andrewthauer replied:
I agree. It's not a problem for us at the moment. Although we have a different primary auth provider then GH, so we need to support 2 auth providers to gain some features via the GH OAuth. It's fine, just not ideal. It's really nice that backstage prompts when GH auth is required though 😄.

andrewthauer reviewed (commented):
🎉 Super excited about this! I realize this is a WIP, but I left some comments and notes anyhow.

mcalus3 commented on plugins/catalog-backend/src/ingestion/LocationAnalyzer.ts:
```diff
@@ -0,0 +1,50 @@
+  async generateConfig(
+    request: AnalyzeLocationRequest,
+  ): Promise<AnalyzeLocationResponse> {
+    const [ownerName, repoName] = request.location.target.split('/').slice(-2);
```
[git-url-parse](https://github.com/coala/git-url-parse) library?

andrewthauer replied:
https://github.com/IonicaBizau/git-url-parse which is already used throughout the project In various places.

mcalus3 replied:
Fixed

mcalus3 commented on plugins/catalog-backend/src/ingestion/LocationAnalyzer.ts:
```diff
@@ -0,0 +1,50 @@
+        name: repoName,
+        annotations: { 'github.com/project-slug': `${ownerName}/${repoName}` },
+      },
+      spec: { type: 'service', owner: ownerName, lifecycle: 'experimental' },
```
I'll do it directly first, optionally in the future splitting it to separate processor. If _unknown_ lifecycle and _other_ type won't cause any problems in usage then It's better to use it.

andrewthauer replied:
Using "unknown" shouldn't cause any issues as the type field can be anything and the common terms (eg "experimental") are purely default recommendations in the entity spec.

Agreed, about deferring the custom processor. It was more a general idea. I will take a look at ours and see if it's possible to refactor it into something more generalized. It could come in handy for this along with the more traditional ingestion flow.

mcalus3 replied:
Fixed

mcalus3 commented on plugins/catalog-backend/src/ingestion/types.ts:
```diff
@@ -68,3 +69,69 @@ export type ReadLocationError = {
+  field: string;
+
+  // The outcome of the analysis for this particular field
+  state:
```
So regular camelCase then? I too think it's better.

mcalus3 replied:
Fixed

mcalus3 commented on plugins/catalog-backend/src/service/router.ts:
```diff
@@ -33,13 +36,19 @@ export interface RouterOptions {
+    entitiesCatalog,
+    locationsCatalog,
+    higherOrderOperation,
+    locationAnalyzer: locationAnalyzer,
```
It's leftover after rename, will delete it.

mcalus3 replied:
Fixed

mcalus3 commented on plugins/catalog-import/src/api/CatalogImportApi.ts:
```diff
@@ -0,0 +1,39 @@
+
+export interface CatalogImportApi {
+  submitPRToRepo(options: {
+    token: string;
```
It's an OAUTH token issued for the logged GitHub user. Do you think this argument should be named _oauthToken_ instead?

andrewthauer replied:
Might be more clear. However, is the intention to have this work with other Git providers in the future? If so then I'm not sure it should be part of the interface then. Is it possible to have this only in the API implementation?

mcalus3 replied:
Name changed

mcalus3 commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -0,0 +1,188 @@
+import { AnalyzeLocationResponse } from '@backstage/plugin-catalog-backend';
+import { PartialEntity } from '../util/types';
+
+export const API_BASE_URL = '/api/catalog/locations';
```
It is using it already, this const should be deleted.

mcalus3 replied:
Fixed

mcalus3 commented on plugins/catalog-backend/src/ingestion/LocationAnalyzer.ts:
```diff
@@ -0,0 +1,50 @@
+      kind: 'Component',
+      metadata: {
+        name: repoName,
+        annotations: { 'github.com/project-slug': `${ownerName}/${repoName}` },
```
Fixed

mcalus3 commented on plugins/catalog-import/src/api/CatalogImportApi.ts:
```diff
@@ -0,0 +1,39 @@
+});
+
+export interface CatalogImportApi {
+  submitPRToRepo(options: {
```
Fixed

mcalus3 commented on plugins/catalog-import/src/components/ComponentConfigDisplay.tsx:
```diff
@@ -0,0 +1,89 @@
+type Props = {
+  nextStep: () => void;
+  configFile: ConfigSpec;
+  savePRLink: (PRLink: string) => void;
```
Fixed

mcalus3 commented on plugins/catalog-import/tsconfig.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Deleted

mcalus3 commented on plugins/catalog-import/src/components/ImportFinished.tsx:
```diff
@@ -0,0 +1,66 @@
+  PRLink: string;
+};
+
+export const ImportFinished: React.FC<Props> = ({ nextStep, PRLink }) => {
```
Fixed

andrewthauer commented on plugins/catalog-backend/src/ingestion/LocationAnalyzer.ts:
```diff
@@ -0,0 +1,52 @@
+ */
+
+import { Logger } from 'winston';
+import GitUriParser from 'git-url-parse';
```
Since this is a function and not a class it should probably be lowercase. The existing convention in the code base is like https://github.com/backstage/backstage/blob/master/packages/backend-common/src/reading/BitbucketUrlReader.ts#L18

mcalus3 replied:
But I've copied it from other place in the codebase:
https://github.com/backstage/backstage/blob/0c0798f082ae6f443798196d835d727837046477/plugins/scaffolder-backend/src/scaffolder/stages/prepare/github.ts#L23
(will fix with next commit)

andrewthauer replied:
It looks like the scaffolder is the only spot not following the lowercase function convention used elsewhere :(

freben replied:
This is one reason that we have https://github.com/backstage/backstage/blob/master/docs/architecture-decisions/adr003-avoid-default-exports.md :) If you search for `from 'git-url-parse'` you'll see that there are a bunch of different names chosen. But `parseGitUrl` is by far the most common and it makes sense to me to be a verb in the imperative form since it's a function that performs an action.

mcalus3 commented:
@andrewthauer @freben 

I've implemented a flow from this mockup @dtuite was posting here a week ago (https://whimsical.com/github-importer-UiLo6FicL54kw2Do1KxDbY).

Here is a video of it:
https://www.loom.com/share/318d68452c4549539751b00356d42ab8

I've replaced a `register-component` plugin so if it's the right direction in the architecture, it can be removed completely as it's not used right now. I've incorporated all it's features into one of the flows of the `catalog-import` plugin. Right now, components are 'if-heavy' but if in the future (i.e. if some third 'target type' would be added), this would be easily refactored to the architecture in which different per-type react components are being injected to further steps of the stepper depending on what url user pastes in the first step.

Rugvip commented on plugins/catalog-backend/src/service/router.ts:
```diff
@@ -131,6 +140,14 @@ export async function createRouter(
+    router.post('/analyze-location', async (req, res) => {
+      const input = await validateRequestBody(req, analyzeLocationSchema);
+      const output = await locationAnalyzer.generateConfig(input);
+      res.status(201).send(output);
```
Just a `200` right? Nothing actually being created and we're not returning a `Location` header

mcalus3 replied:
Done

Rugvip commented on plugins/catalog-import/package.json:
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "@backstage/plugin-catalog-import",
+  "version": "0.1.1-alpha.26",
```
Remember to bump versions here :grin:

mcalus3 replied:
Bumped to 0.2.0 along with all backstage dependencies. I have a problem merging this branch to master though...

Rugvip commented on plugins/catalog-import/src/api/CatalogImportApi.ts:
```diff
@@ -0,0 +1,36 @@
+
+export interface CatalogImportApi {
+  submitPrToRepo(options: {
+    oAuthToken: string;
```
Better have this happen internally in the implementation instead by depending on the appropriate auth APIs. This is hardcoded for GitHub anyway so might as well depend on GitHub auth directly. Having that work a bit more dynamically based on what type of repo we're working with is something we need to figure out in general anyway.

mcalus3 replied:
How to use GitHub auth without the `useApi` hook? Can some GithubApi client be injected to the constructor of my `CatalogImportApi`?

mcalus3 replied:
Well, I've done this exactly :)

mcalus3 commented:
@freben @stefanalund I've resolved conflicts with master branch and bumped package versions, so the PR is ready to be merged :)

andrewthauer reviewed (approved):
:shipit: and iterate on anything additional?

benjdlambert commented:
Hey @mcalus3 - looks like the build here is still failing, think you'll need to update the dependencies in your plugin to point at the latest versions that we shipped yesterday

Rugvip commented on plugins/catalog-backend/src/ingestion/types.ts:
```diff
@@ -68,3 +69,69 @@ export type ReadLocationError = {
+};
+
+export type AnalyzeLocationResponse = {
+  existingEntityFiles: AnalyzeLocationExistingEntity[];
```
Wonder if we could skip this part of the interface for now, since it's not used?

mcalus3 replied:
These types were suggested by @freben so I pasted them all to have some reminder what was intended for the future for this feature. Of course I can remove them, but does it make any problem if they are left out here?

freben replied:
I vote keep :) We talked before about the flow where you just wanna register a url and we go "oh there's already a file in there, let's make sure there's a location for it"

Rugvip commented on plugins/catalog-backend/src/ingestion/types.ts:
```diff
@@ -68,3 +69,69 @@ export type ReadLocationError = {
+// LocationAnalyzer
+//
+
+export type LocationAnalyzer = {
```
Should give this one a nice JsDoc comment, what's the purpose of it etc.

Using JsDoc/TsDoc (`/** */`) comments will make the docs show up in editors as well, which is 👌

mcalus3 replied:
Added, although quite minimal

Rugvip commented on plugins/catalog-backend/src/ingestion/types.ts:
```diff
@@ -68,3 +69,69 @@ export type ReadLocationError = {
+//
+
+export type LocationAnalyzer = {
+  generateConfig(
```
Would probably call this `analyzeLocation` instead?

mcalus3 replied:
Renamed.

Rugvip commented on plugins/catalog-backend/src/ingestion/types.ts:
```diff
@@ -68,3 +69,69 @@ export type ReadLocationError = {
+// We should probably have a chat about how this can best be conveyed to
+// the frontend. It'll probably contain a (possibly incomplete) entity, plus
+// enough info for the frontend to know what form data to show to the user
+// for overriding/completing the info.
```
As @freben has mentioned in discussions it could make sense to transform this into JSON Schema, with existing values, missing fields, etc. I'd really like to keep that kind of logic outside of the catalog backend itself though, keeping the additions for this use-case to the bare minimum needed to support it and instead having a separate `catalog-import-backend` plugin or so that does most of the work.

My hunch here is that it would essentially be enough to be able to dry-run catalog processing without the final validation step, and then use the output entities and the base for that schema by inspecting them for missing fields etc.

Rugvip commented on plugins/catalog-backend/src/ingestion/LocationAnalyzer.ts:
```diff
@@ -0,0 +1,52 @@
+  async generateConfig(
+    request: AnalyzeLocationRequest,
+  ): Promise<AnalyzeLocationResponse> {
+    const { owner, name, source } = parseGitUri(request.location.target);
```
I think we could skip the owner here, it really just makes sense for user repos or specific org structures. An org that always has the github org == owner can instead have a catalog processor for that which will end up populating our entity anyway.

mcalus3 replied:
Removed owner from generated config, although it's still used in github repo path generation.

Rugvip commented on plugins/catalog-backend/src/ingestion/LocationAnalyzer.ts:
```diff
@@ -0,0 +1,52 @@
+  LocationAnalyzer,
+} from './types';
+
+export class LocationAnalyzerClient implements LocationAnalyzer {
```
Should rename this as it isn't a client :grin:, `RepoLocationAnalyzer` perhaps?

mcalus3 replied:
Renamed.

Rugvip reviewed (commented):
Awesome work, thanks! :grin:

Couple of smaller renames/cleanups that I think would make sense to get in before we merge. My comments about the architecture is definitely something that should wait, but I do think we'll want to extract much of this logic into a separate backend plugin in the future. That'll let us evolve and add features to the import process more quickly, without caring as much whether it bloats and fits with the long-term vision for the catalog.

freben commented on plugins/catalog-import/src/api/CatalogImportApi.ts:
```diff
@@ -0,0 +1,35 @@
+    fileContent: string;
+  }): Promise<{ link: string; location: string }>;
+  createRepositoryLocation(options: { location: string }): Promise<void>;
+  generateEntityDefinitions(options: {
```
At a glance it's not clear to me what these two methods actually do. For example for the second one I would have expected it to return something more similar to what the backend returns? I'm probably misunderstanding. Maybe add jsdoc comments to them.

freben commented on plugins/catalog-import/src/api/CatalogImportApi.ts:
```diff
@@ -0,0 +1,35 @@
+    owner: string;
+    repo: string;
+    fileContent: string;
+  }): Promise<{ link: string; location: string }>;
```
What I'm getting at is that there will not be just one of these instantiated at any given time. There will probably be N instances, one for each provider, and then we select among them based on the location URL or something like that.

freben commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -0,0 +1,192 @@
+        },
+        method: 'POST',
+        body: JSON.stringify({
+          location: { type: 'github', target: repo },
```
same here, needs to be pluggable - should really be using the `url` type, right?

freben commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -0,0 +1,192 @@
+        },
+        method: 'POST',
+        body: JSON.stringify({
+          type: 'github',
```
should be url as well, right?

freben commented on plugins/catalog-import/src/api/CatalogImportClient.ts:
```diff
@@ -0,0 +1,192 @@
+    repo: string;
+    fileContent: string;
+  }): Promise<{ link: string; location: string }> {
+    const token = await this.githubAuthApi.getAccessToken(['repo']);
```
We had a chat about this today, how the auth api is not very suitable at this point for this purpose. You'd probably want an api that's shaped a bit differently, where providers can be listed and auth acquired for a given one of those providers etc. Well, future work.

freben commented on plugins/catalog-import/src/components/ComponentConfigDisplay.tsx:
```diff
@@ -0,0 +1,173 @@
+
+  return (
+    <Grid container direction="column" spacing={1}>
+      {!parseGitUri(configFile.location).filepathtype ? (
```
this boolean keeps popping up over and over again in the file, can you just store it in a named variable at the top so it's easier to know what it means and how it's used?

freben commented on plugins/catalog-import/src/components/ImportComponentForm.tsx:
```diff
@@ -0,0 +1,131 @@
+          variant="outlined"
+          label="Repository URL"
+          error={hasErrors}
+          placeholder="https://github.com/spotify/backstage"
```
it's backstage/backstage now :)

freben commented on plugins/catalog-import/src/components/ImportComponentPage.tsx:
```diff
@@ -0,0 +1,111 @@
+      <Content>
+        <ContentHeader title="Start tracking your component on backstage">
+          <SupportButton>
+            Start tracking your component in Backstage. TODO: Add more
```
the todo could be an html (or jsx `{/* */}`) comment instead so it isn't visible to users

freben reviewed (approved):
a number of comments - the only one i think should be done first is to add a changeset file, the rest can happily be done after merge

mcalus3 commented on plugins/catalog-backend/src/ingestion/types.ts:
```diff
@@ -68,3 +69,69 @@ export type ReadLocationError = {
+
+export type LocationAnalyzer = {
+  generateConfig(
+    location: AnalyzeLocationRequest,
```
it's now `analyzeLocation`

mcalus3 commented on plugins/catalog-backend/src/ingestion/types.ts:
```diff
@@ -68,3 +69,69 @@ export type ReadLocationError = {
+
+  // If the analysis did suggest a value, this is where it would be. Not sure if we want
+  // to limit this to strings or if we want it to be any JsonValue
+  value: string | null;
```
I always treated it as:
undefined - unintentional lack of value (property not set)
null - intentional lack of value (set as null until the value comes from the promise, handle null on the frontend as "the value is not there yet")

mcalus3 commented on plugins/catalog-backend/src/service/CatalogBuilder.ts:
```diff
@@ -229,11 +232,13 @@ export class CatalogBuilder {
       entitiesCatalog,
       locationsCatalog,
       higherOrderOperation,
+      locationAnalyzer,
```
Done, you can review if it's what you meant.

freben commented on .changeset/smart-turkeys-bathe.md:
```diff
@@ -0,0 +1,30 @@
+---
+'@backstage/plugin-catalog-backend': minor
+'@backstage/plugin-catalog-import': minor
```
missing backend-common here which is the big one that did get the change

mcalus3 replied:
I didn't see any files changed in the packages/backend-common directory, I've selected only plugins/packages which got files changed.

freben replied:
Ah yes you are quite right, sorry i completely mixed it up, of course the builder is in plugin-catalog-backend.

Rugvip commented on .changeset/smart-turkeys-bathe.md:
```diff
@@ -0,0 +1,30 @@
+---
+'@backstage/plugin-catalog-backend': minor
```
Would just be a patch bump here right? Don't really need an entry for the new plugin either as that'll just result in a bump in the next release without any changes.

mcalus3 replied:
@freben thinks not. It's minor bump because we add new feature to it - location analyzer.

Rugvip reviewed (approved):
:shipit:

dtuite commented:
Are there comments still left to address or is this ready to go?

dtuite commented:
I just noticed that this is logging errors and stack traces when used. It's correct that the `catalog-info.yaml` file lookup is a 404 because I haven't merged the PR yet but I don't think it makes sense to log an error and stack trace since this is expected.

```
2020-11-21T18:31:01.872Z catalog warn Failed item in location github:https://github.com/RoadieHQ/test-data-api/blob/main/catalog-info.yaml, NotFoundError: Unable to read github, NotFoundError: https://github.com/RoadieHQ/test-data-api/blob/main/catalog-info.yaml could not be read as https://api.github.com/repos/RoadieHQ/test-data-api/contents/catalog-info.yaml?ref=main, 404 Not Found
[1]     at Module.notFoundError (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/processors/results.ts:40:12)
[1]     at UrlReaderProcessor.readLocation (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/processors/UrlReaderProcessor.ts:72:70)
[1]     at runMicrotasks (<anonymous>)
[1]     at processTicksAndRejections (internal/process/task_queues.js:93:5)
[1]     at async LocationReaders.handleLocation (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/LocationReaders.ts:142:13)
[1]     at async LocationReaders.read (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/LocationReaders.ts:74:11)
[1]     at async HigherOrderOperations.refreshSingleLocation (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/HigherOrderOperations.ts:153:26)
[1]     at async HigherOrderOperations.refreshAllLocations (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/HigherOrderOperations.ts:134:9)
[1]     at async startRefresh (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/util/runPeriodically.ts:42:9)
```

dtuite reviewed (changes requested):
I did some local testing. I don't seem to end up with a component in the catalog once I merge the PR which this feature opens.

I also noticed some warnings in the logs which probably shouldn't be there.

I will check with Marek on Monday.

dtuite commented on plugins/catalog-import/src/components/ComponentConfigDisplay.tsx:
```diff
@@ -0,0 +1,173 @@
+        <Typography>
+          Following config object will be submitted in a pull request to the
+          repository{' '}
+          <Link href={configFile.location}>{configFile.location}</Link> and
```
This should have `target="_blank" rel="noopener noreferrer"` added to prevent stealing the tab while the user is in the middle of an import process.

mcalus3 replied:
Done

benjdlambert commented:
Should we merge without @dtuite changes and maybe create a separate ticket to track that?

dtuite commented:
@benjdlambert I'd suggest just holding off for a little bit. @mcalus3 is investigating as we speak to see how involved the fix might be. If it's quick, we can get it in.

The errors and stack traces in the logs are the worst part IMO. If you're ok with merging it then I guess I am.

mcalus3 commented on plugins/catalog-backend/src/ingestion/types.ts:
```diff
@@ -68,3 +69,75 @@ export type ReadLocationError = {
+  // "Found a CODEOWNERS file that covers this target, so we suggest leaving this
+  // field empty; which would currently make it owned by X" where X is taken from the
+  // codeowners file.
+  description: string;
```
In the `CodeOwnersProcessor` there is a function which throws an error if the codeowners file is not present, so I assume this is deliberate that these warning are being logged. It's severity level is warning so it should be ok. It's definitely not related to this pull request (it started to occur now because I recently removed the `owner` field from generated configs and CodeOwnersProcessor isn't even invoked if owners field in the configuration is present)

andrewthauer replied:
Yes, that makes sense. If the `owner` field is already present then the `CodeOwnersProcessor` will not attempt to fetch the owner via a `CODEOWNERS` file. This brings up an interesting question though. Some organizations (like ours) use the `CODEOWNERS` extensively likely and others do not. So if this plugin does not include an `owner` it will be an ideal experience for some folks but will actually not work end-to-end for others as the `owner` is a required field.

freben replied:
The idea in the end is that the import will be smart enough to be able to say "i found codeowners so I'll recommend leaving this empty (but feel free to override me)", vs "i was not able to find a sensible user so you'll have to fill one in in a form in the frontend before posting submit "

mcalus3 commented on plugins/catalog-import/src/components/ComponentConfigDisplay.tsx:
```diff
@@ -0,0 +1,173 @@
+          >
+            Next
+          </Button>
+        </Grid>
```
Added:
https://www.loom.com/share/3318602cf2ed4f06904815f3515eb224

mcalus3 commented on plugins/catalog-import/src/components/ImportFinished.tsx:
```diff
@@ -0,0 +1,53 @@
+      </Grid>
+      <Grid item>
+        {type === 'repo' ? (
+          <Link href={PRLink} style={{ marginRight: '8px' }}>
```
Done

mcalus3 commented:
> 
> 
> I just noticed that this is logging errors and stack traces when used. It's correct that the `catalog-info.yaml` file lookup is a 404 because I haven't merged the PR yet but I don't think it makes sense to log an error and stack trace since this is expected.
> 
> ```
> 2020-11-21T18:31:01.872Z catalog warn Failed item in location github:https://github.com/RoadieHQ/test-data-api/blob/main/catalog-info.yaml, NotFoundError: Unable to read github, NotFoundError: https://github.com/RoadieHQ/test-data-api/blob/main/catalog-info.yaml could not be read as https://api.github.com/repos/RoadieHQ/test-data-api/contents/catalog-info.yaml?ref=main, 404 Not Found
> [1]     at Module.notFoundError (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/processors/results.ts:40:12)
> [1]     at UrlReaderProcessor.readLocation (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/processors/UrlReaderProcessor.ts:72:70)
> [1]     at runMicrotasks (<anonymous>)
> [1]     at processTicksAndRejections (internal/process/task_queues.js:93:5)
> [1]     at async LocationReaders.handleLocation (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/LocationReaders.ts:142:13)
> [1]     at async LocationReaders.read (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/LocationReaders.ts:74:11)
> [1]     at async HigherOrderOperations.refreshSingleLocation (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/HigherOrderOperations.ts:153:26)
> [1]     at async HigherOrderOperations.refreshAllLocations (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/ingestion/HigherOrderOperations.ts:134:9)
> [1]     at async startRefresh (webpack-internal:///../../node_modules/@backstage/plugin-catalog-backend/src/util/runPeriodically.ts:42:9)
> ```

This has to be done in the `UrlReaderProcessor` (https://github.com/backstage/backstage/blob/0d58066ee5a6bb301cd43e186944858fae64164f/plugins/catalog-backend/src/ingestion/processors/UrlReaderProcessor.ts#L42) but it's a good for separate issue because I'm not sure if errorshould be completely supressed of only changed to some "info log" that the location is not present there yet. I also don't know how this relates to the "optional" argument to that `readLocation` method.

mcalus3 commented:
@benjdlambert It can be merged now, @dtuite comments are addressed.

benjdlambert commented on .changeset/smart-turkeys-bathe.md:
```diff
@@ -0,0 +1,30 @@
+---
+'@backstage/plugin-catalog-backend': minor
+'@backstage/plugin-catalog-import': minor
+'example-app': patch
```
don't think you need to bump either this or the `example-backend`

mcalus3 replied:
You think that I shouldn't bump patch versions of example-app and example-backend? I've modified files in `packages/app` and `packages/backend` directories so thought that it means they need a version bump.

freben replied:
Not needed - they are private packages that aren't released; they are essentially like unversioned

mcalus3 replied:
I've changed it so example-app and example-backend are not updated

Rugvip commented on plugins/catalog-backend/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@backstage/plugin-catalog-backend",
-  "version": "0.2.2",
+  "version": "0.3.0",
```
These version bumps shouldn't be included, they happen automatically based on the changesets in the release PR

mcalus3 replied:
I've reverted `catalog-backend` version to 0.2.2 and reverted it's dependency versions to 0.2.2 in `example-backstage` and `catalog-import` packages

Rugvip commented on yarn.lock:
```diff
@@ -1318,6 +1320,37 @@
 
+"@backstage/plugin-catalog-backend@^0.2.2":
+  version "0.2.2"
+  resolved "https://registry.npmjs.org/@backstage/plugin-catalog-backend/-/plugin-catalog-backend-0.2.2.tgz#4e17e0376a920f1425de5b0997d34134ff101a11"
```
Needs a `yarn install` run :grin:

mcalus3 replied:
Done

Rugvip commented on yarn.lock:
```diff
@@ -1308,7 +1308,9 @@
 "@backstage/catalog-model@^0.2.0":
-  version "0.3.0"
+  version "0.2.0"
+  resolved "https://registry.npmjs.org/@backstage/catalog-model/-/catalog-model-0.2.0.tgz#e3fe2a4ddeb6a9b6ec480c80cb2b9c39cb245576"
```
This shouldn't have an entry in the lockfile either, fixed by switching the catalog-model dep to 0.3.0

freben commented:
Merged!

As the build status mentioned, the codecov level was just 11% and it lowered the overall project coverage a bit which contributes to risk for future PRs to hit the "global" limit. We'd appreciate if you took some time to write tests to bump up the coverage for these additions.

mcalus3 commented:
> 
> 
> Merged!
> 
> As the build status mentioned, the codecov level was just 11% and it lowered the overall project coverage a bit which contributes to risk for future PRs to hit the "global" limit. We'd appreciate if you took some time to write tests to bump up the coverage for these additions.

I'll try to do that in this week and open a subsequent PR
